rfcs/0004-insecure-http/README.md

```diff
@@ -0,0 +1,209 @@
+Some Flux objects provide a `.spec.insecure` field to enable the use of non-TLS based endpoints, but they don't clearly notify
+users when the option is not supported (e.g. Azure/GCP Buckets).
+
+### Goals
+* Provide a flag across relevant Flux controllers which disables all outgoing HTTP connections.
+* Add a field which enables the use of non-TLS endpoints to appropriate Flux objects.
+* Provide a way for users to be made aware that their use of non-TLS endpoints is not supported if that is the case.
+
+### Non-Goals
+* Break Flux's current behavior of allowing HTTP connections.
+* Change in behavior of communication between Flux components.
+
+## Proposal
+
+### Controllers
+Flux users should be able to enforce that controllers are using HTTPS connections only.
+This shall be enabled by adding a new boolean flag `--insecure-allow-http` to the following controllers:
+* source-controller
+* notification-controller
+* image-automation-controller
+* image-reflector-controller
+
+The default value of this flag shall be `true`. This would ensure that there is no breaking change with controllers
+still being able to access non-TLS endpoints. To disable this behavior and enforce the use of HTTPS connections, users would
+have to explicitly pass the flag to the controller:
+
+```yaml
+spec:
+  template:
+    spec:
+      containers:
+      - name: manager
+        image: fluxcd/source-controller
+        args:
+          - --watch-all-namespaces
+          - --log-level=info
+          - --log-encoding=json
+          - --enable-leader-election
+          - --storage-path=/data
+          - --storage-adv-addr=source-controller.$(RUNTIME_NAMESPACE).svc.cluster.local.
+          - --insecure-allow-http=false
+```
+
+> Note: The flag shall not be added to the following controllers:
+> * kustomize-controller: This flag is excluded from this controller, as the upstream `kubenetes-sigs/kustomize` project
+> does not support disabling HTTP connections while fetching resources from remote bases. We can revisit this if the
+> upstream project adds support for this at a later point in time.
+> * helm-controller: This flag does not serve a purpose in this controller, as the controller does not make any HTTP calls.
+> Furthermore although both controllers can also do remote applies, serving `kube-apiserver` over plain
+> HTTP is disabled by default. While technically this can be enabled, the option for this configuration was also disabled
+> quite a while back (ref: https://github.com/kubernetes/kubernetes/pull/65830/).
+
+Both kustomize-controller and helm-controller currently have a flag `--insecure-kubeconfig-tls` which makes the controller skip
+TLS verification when connecting to a Kubernetes cluster with an HTTPS connection. This flag shall be renamed to
+`--insecure-skip-tls-verify` to align it with the Flux CLI which offers this command for the same purpose.
```

Comment:
Does this belong in this RFC? Skip TLS verification in block insecure HTTP connection RFC. Seems unrelated and may deserve its own RFC or dealt separately.

Reply:
The main reason why I suggested we covered this here was to take into account cases in which the scheme does not highlight the use of TLS (e.g. `oci://`).  In such cases, the use of the field `spec.insecure=True` would be blocked off when `--insecure-skip-tls-verify` is set to `false`.

The `--insecure-allow-http` would only apply and be relevant when the schemes `http://` and `https://` as being used.

Reply:
> The --insecure-allow-http would only apply and be relevant when the schemes http:// and https:// as being used.

This is so confusing, are you suggesting that to block all insecure connections, an admin will have to set 2 flags in source-controller: `--insecure-allow-http` and `--insecure-skip-tls-verify` ?

Reply:
They control separate things: `--insecure-allow-http` allows connections to HTTP endpoints, where as `--insecure-skip-tls-verify` skips the verification of TLS certificates over HTTPS (but does not allow HTTP).

If `--insecure-allow-http` is thus set, HTTP and HTTPS connections can be made but a HTTPS connection would still require a valid TLS certificate. While if `--insecure-skip-tls-verify` is set but `--insecure-allow-http` is not, HTTPS connections are required but the TLS verification is skipped.

Reply:
@hiddeco but that's not what @pjbgf is saying:

> the use of the field spec.insecure=True would be blocked off when --insecure-skip-tls-verify is set to false.

The use of `OCIRepository.spec.insecure` is for when the OCI repo is hosted on a plain HTTP non-TLS registry.

Reply:
That seems wrong (probably due to a misunderstanding of what the field enables), I think the behavior as I described is correct. Skipping TLS verify for at least Helm OCI does not even seem to be an option at present.

Reply:
My suggestion aligns with what @hiddeco mentioned. `--insecure-allow-http` is for the use of the `http://` scheme from a user-perspective. 

For OCI and Buckets sources, TLS assurances are not scheme bound - at least not from a user PoV. Therefore, the only ability they have to enforce such controls is via `spec.Insecure`. Blocking such use would be via `--insecure-skip-tls-verify`.

Reply:
@pjbgf IMO to block OCIRepositories when `spec.insecure=True` admins will set `--insecure-allow-http`, to block OCIRepositories when `spec.certSecretRef`  admins will set `--insecure-skip-tls-verify`.

Reply:
I fail to see the relationship between `--insecure-skip-tls-verify` and `spec.insecure`. `spec.insecure` is solely for configuring whether the connection uses HTTP or HTTPS. `--insecure-skip-tls-verify` is only considered when the connection being used is an HTTPS one in the first place.
If `spec.insecure=true`,  then the value of `--insecure-skip-tls-verify` is inconsequential automatically because the connection is going to be made using the HTTP protocol, hence there's no TLS involved.
To further clarify any future questions, unlike configuring connections to be HTTP or HTTPS, there is no way to skip TLS verification at an object level, its only possible via the flag `--insecure-skip-tls-verify`.

Reply:
I was under the impression that `spec.insecure` in Buckets skipped TLS verification. After going through the minio code I realised that that wasn't the case, but it rather just adjust TLSConfig settings, therefore I concur with:
- Removal of `--insecure-skip-tls-verify` from this RFC.
- Use of `--insecure-allow-http` to block `spec.Insecure=True` and the use of `http://`.